### ISSUE TITLE
Injecting services within HandleProxyRequest delegate

### ISSUE THREAD

**Issue by ben-foster-cko** [damianh/ProxyKit#117]
Hi,

I'd like to implement a proxy that supports templated request content and proxies the rendered template to the upstream services. A simplified example is:

```c#
            app.RunProxy(async context =>
            {
                string requestBody = await ReadRequestBodyAsync(context.Request);
                
                ForwardContext forwardContext = context
                    .ForwardTo("https://upstream")
                    .AddXForwardedHeaders();

                string rendered = await BindAsync(requestBody, new SomeModel
                {
                    Name = "Ben Foster"
                });

                forwardContext.UpstreamRequest.Content = new StringContent(rendered, Encoding.UTF8, context.Request.ContentType);

                return await forwardContext.Send();
            });
```

In the above example I'm hard-coding the `SomeModel` object but this would actually need to be loaded from a DB based on a value in one of the request headers.

What would the recommended approach be for injecting services into the delegate. I've considered:

1. Using `context.RequestServices`
2. Using another middleware component to load the required data and adds it to `HttpContext.Items`

**Comment by damianh** [comment]
Right now the to get a request scoped service is `context.RequestServices`. While `HttpConext.Items` would work but it feels less desirable. 

But I get how that might feel a bit service-locator-y.

💭 off top of my head perhaps being able to define a class to which you could inject a service would be useful.... I'll spike an idea...

**Comment by ben-foster-cko** [comment]
Yeah I went with 1. in the end. Would be nice if it worked similar to standard ASP.NET Core middleware  i.e. define a class with an Invoke method. Otherwise, could always derive from `ProxyMiddleware`?

**Comment by damianh** [comment]
I've spiked a "Typed Handler" over at #119. Could you go over at take a look?

**Comment by damianh** [comment]
Fixed via of #119.  Services can be injected into a typed handler. Request scoped services will still need to come from `HttpContext.RequestServices`
